server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,124 @@
+
+    public static final Setting<Settings> THRESHOLD_SETTINGS =
+            Setting.groupSetting("master.throttling.thresholds.", Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    /**
+     * To configure more task for throttling, override getMasterThrottlingKey method with task name in task executor.
+     * Verify that throttled tasks would be retry.
+     *
+     * Added retry mechanism in TransportMasterNodeAction so it would be retried for customer generated tasks.
+     */
+    public static Set<String> CONFIGURED_TASK_FOR_THROTTLING =
+        Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+            "update-settings", "cluster-update-settings",
+            "create-index", "auto-create",
+            "delete-index", "delete-dangling-index",
+            "create-data-stream", "remove-data-stream",
+            "rollover-index", "index-aliases",
+            "put-mapping", "refresh-mapping",
+            "close-indices", "open-indices",
+            "create-index-template", "remove-index-template",
+            "create-component-template", "remove-component-template",
+            "create-index-template-v2", "remove-index-template-v2",
+            "put-pipeline", "delete-pipeline",
+            "create-persistent-task", "finish-persistent-task", "remove-persistent-task","update-task-state",
+            "put-script", "delete-script",
+            "put_repository", "delete_repository",
+            "create-snapshot","delete-snapshot","update-snapshot-state",
+            "restore_snapshot",
+            "cluster-reroute-api"
+        )));
+
+    private final int DEFAULT_THRESHOLD_VALUE = -1; // Disabled throttling
+    private final MasterThrottlingStats throttlingStats = new MasterThrottlingStats();
+    private final MasterService masterService;
+
+    public MasterTaskThrottler(final ClusterSettings clusterSettings, final MasterService masterService) {
+        super(true);
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.masterService = masterService;
+    }
+
+    public void validateSetting(final Settings settings) {
+        /**
+         * TODO: Change the version number of check as per version in which this change will be merged.
+         */
+        if(masterService.state().nodes().getMinNodeVersion().compareTo(LegacyESVersion.V_7_10_3) < 0) {
+            throw new IllegalArgumentException("All the nodes in cluster should be on version later than 7.10.3");
+        }
+        Map<String, Settings> groups = settings.getAsGroups();
+        for(String key : groups.keySet()) {
+            if(!CONFIGURED_TASK_FOR_THROTTLING.contains(key)) {
+                throw new IllegalArgumentException("Master task throttling is not configured for given task type: " + key);
+            }
+            int threshold = groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE);
+            if(threshold < DEFAULT_THRESHOLD_VALUE) {
```

Comment:
Why default threshold value? It may change in future? If not, rename default to min?